### PR TITLE
Improve subnav & breadcrumb alignment

### LINKED ISSUE #13462: Visual bug in VPN resource center - Breadcrumb component no longer aligns with nav
## Description

The breadcrumb and nav bar in the VPN resource center no longer align after the nav changes

### Steps to reproduce
Visit https://www.mozilla.org/en-US/products/vpn/resource-center/5-reasons-you-should-use-a-vpn/ (or any other urls in the resource-center)

What should happen?
They should have the same horizontal margins
Screenshot 2023-07-27 at 3 10 37 PM

### LINKED ISSUE #14081: Breadcrumb links on firefox/feature pages are misaligned
## Description

The Protocol breadcrumbs component on the `firefox/feature` pages is misaligned with the rest of the page content.

### Steps to reproduce

Open https://www.mozilla.org/en-US/firefox/features/private/

### Expected result

The breadcrumb links should be aligned with the nav links

image

### Actual result

The breadcrumb is not aligned

image

It could also perhaps do with a little more top / bottom padding, to make the breadcrumb links easier to click?

### Environment

Firefox Nightly
macOS

### PR DESCRIPTION
**Pull request by janbrasna**
## One-line summary

Improves the default breadcrumb styling and adjusts subnav to better align all the navigation components.

## Significant changes and points to review

TL;DR: The subnav & breadcrumbs were aligned as part of the content, not matching the nav masthead above. This aims to reset the content-like styling for these additional nav components, and apply the same layout as the main nav above them.

WIP

not sure i like the looks of "aligned with nav" now on various pages, where the alignment with content underneath looked probably better? 🤷 
 — feel free to browse around locally, i've picked some good examples below in "testing", where are some nice examples of different content alignment, and how that interacts with the repositioned subnav…

- monkey-patched breadcrumb overrides to _sub-navigation.scss as it's compiled into base everywhere important
   - (b/c breadcrumbs are included w/o any styling, implying default protocol, so to override this a "cheap" place ideally loaded everywhere is needed; this just avoids adding a new component and including it in a dozen places…)
- BTW changes VPN breadcrumbs colour for contrast with subnav

## Issue / Bugzilla link

Fixes #14081
Fixes #13462
Supersedes #14253

## Testing

http://localhost:8000/en-US/firefox/features/private/
http://localhost:8000/en-US/firefox/new/
http://localhost:8000/en-US/firefox/enterprise/
http://localhost:8000/en-US/firefox/channel/desktop/
http://localhost:8000/en-US/products/vpn/
http://localhost:8000/en-US/products/vpn/resource-center/is-a-vpn-safe/

### CODE REVIEW

**Comment by reemhamz** [comment]
Hello! I reviewed the PR and have some comments and suggestions, however before I officially write them up, I've posed a question on the issue ticket to help clarify some things before we continue with this change :)

**Comment by janbrasna** [comment]
@reemhamz 👋 I'm actually wondering whether some adjustments like this wouldn't make more sense upstream in Protocol, along with some other alignment improvements in subnav — this has also been already reported for VPN:
- #13462

File: media/css/firefox/features/article.scss [hunk 1]
```diff
@@ -9,6 +9,31 @@ $image-path: '/media/protocol/img';
 @import '~@mozilla-protocol/core/protocol/css/components/callout';
 @import '~@mozilla-protocol/core/protocol/css/components/breadcrumb';
 
+.mzp-c-breadcrumb {
+    padding-top: $spacing-md;
+    padding-bottom: $spacing-md;
+
+    .mzp-c-breadcrumb-item a {
+        color: inherit;
```
Comment (reemhamz) [inline media/css/firefox/features/article.scss:17]: **Nit:** By changing the color of the link, it doesn't make it obvious that it's a link a user can interact with anymore, so it's best to remove this declaration to not make it become an accessibility issue

File: media/css/firefox/features/article.scss [hunk 2]
```diff
@@ -9,6 +9,31 @@ $image-path: '/media/protocol/img';
...
+        .mzp-c-breadcrumb-list {
+            padding: 8px 64px;
+        }
+    }
+
+    @media #{$mq-lg} {
+        .mzp-c-breadcrumb-list {
+            padding: 8px 96px;
+        }
+    }
```
Comment (reemhamz) [inline media/css/firefox/features/article.scss:34]: **Suggestion:** This still doesn't completely fix the misalignment, especially on bigger screens. I think a simple solution would be using a `max-width` property with a `margin: 0 auto` declaration:
```suggestion
        .mzp-c-breadcrumb-list {
              max-width: $content-max;
              margin: 0 auto;
        }
```

This essentially is just copying what the site header has in place, and both should successfully align now :)

File: media/css/firefox/features/article.scss [hunk 3]
```diff
@@ -9,6 +9,31 @@ $image-path: '/media/protocol/img';
 @import '~@mozilla-protocol/core/protocol/css/components/callout';
 @import '~@mozilla-protocol/core/protocol/css/components/breadcrumb';
 
+.mzp-c-breadcrumb {
+    padding-top: $spacing-md;
+    padding-bottom: $spacing-md;
```
Comment (reemhamz) [inline media/css/firefox/features/article.scss:14]: **Suggestion:** This change would probably be better placed in the actual Protocol design system rather than it just being in Bedrock?

**Review by reemhamz — CHANGES_REQUESTED** [review]
Nice work on this! 

As Alex said in the issue, this seems to be a fix for Bedrock because the subnav isn't backported into Protocol 
I've left a few comments that need to be addressed. I also have a strong suggestion:

Since this Breadcrumbs component is used in different pages, and will be used in other pages in the future, we should make it into a CSS bundle that can easily be imported to other pages using this component, especially because we now have some new CSS that needs to align with the website's main nav.

Here's more information on how it works: https://bedrock.readthedocs.io/en/latest/coding.html#id2

1. So, we could move this Breadcrumb-specific CSS into its own file here: https://github.com/mozilla/bedrock/tree/main/media/css/protocol/components -- make sure to import the following as well:

```
@import '~@mozilla-protocol/core/protocol/css/includes/lib';
@import '~@mozilla-protocol/core/protocol/css/components/breadcrumb';
```

2. Then create a CSS bundle for the component in `static-bundles.json` (make sure to restart your server after doing this so the file updates)

3. Import the CSS bundle in the HTML pages using the component
4. You can now remove `@import '~@mozilla-protocol/core/protocol/css/components/breadcrumb';` from relative CSS files


Let me know if you need any help with this!

File: media/css/firefox/features/article.scss (hunk 2 above)
Comment (janbrasna) [reply media/css/firefox/features/article.scss:34]: Exactly, I've noticed that every viewport size has its own alignment issues so I'll try to use the same logic the masthead uses to match it not just "close enough", but reliably in all sizes.

**Comment by alexgibson** [comment]
> Since this Breadcrumbs component is used in different pages, and will be used in other pages in the future, we should make it into a CSS bundle that can easily be imported to other pages using this component, especially because we now have some new CSS that needs to align with the website's main nav.

Since so many pages use a sub nav, I wonder if it might actually be easier to add these styles to the main Protocol base stylesheet, so they are kept along with the main nav styles? That might make for easier maintenance and saves on an HTTP request.

**Comment by janbrasna** [comment]
@alexgibson Yup I'm looking into this, basically since this will be touching the subnav that lives in `css/protocol/components/_sub-navigation` anyways I'll try to accommodate the breadcrumb "local adjustments" in similar fashion, it shouldn't be more than a few lines. If you have a good place in mind (existing file, or a preference to create a new component), I'm open to suggestions. I'm not monkey-patching the `.c-breadcrumbs` to the bottom of `.c-menu` or `.c-navigation` without prior approval ;D

**Comment by alexgibson** [comment]
If it's only a few lines of CSS then perhaps adding it to `css/protocol/components/_sub-navigation` makes sense, but I'm open to @reemhamz thoughts here as well, as this was just a passing thought.

**Comment by reemhamz** [comment]
That can also work @alexgibson! I see the benefit of saving on HTTP requests + it's one less step for developers needing to implement the Breadcrumb component on a page 👍🏼

File: media/css/firefox/features/article.scss (hunk 1 above)
Comment (janbrasna) [reply media/css/firefox/features/article.scss]: Thanks @reemhamz but I can't take any credit for this 😇 it's just repurposing the code `@nathan-barrett` originally wrote for `stories/articles`, and I didn't want it to disappear in git history now that the content is gone;)

But you're right, the styling should be more generic, and also cover all the occurrences of the breadcrumbs (VPN articles is a good example where the bad alignment shows best) so I'm backing out of this local change, and starting over together with fixes to subnav, basically treating the subnav+breadcrumbs containers as close as possible to the masthead, not the content defaults they inherit.

Thanks for the hand-holding, I really appreciate it! Sure I've read about the bundles needing restarts but have forgotten about it since;) so better to get reminded a few times than scratching my head why can't get it loaded;)

I'm trying to improve things that I feel comfortable with, either from what I can infer from some historic patterns, or current epics — as not to bother anyone else too much, having pretty good idea what all you have to juggle… — so you can focus on what _needs_ to be done, and I'm toying with random cosmetic stuff that I can either reliably resolve, or bail out fast if any broader involvement is needed. So I'll try to do my best to only bring up important decisions, not bothering you with me tripping over things, that's for me to figure out… (let's call it a learning curve, _one pebble at a time_, right?)

I'll see if I can apply fixes without introducing any new bundles (as subnav is included everywhere important, in base; and historically breadcrumbs never needed any custom styles so I'd be awkward for everyone to get used to including a bundle for them everywhere now…) and once the layout is something folks would like to see happening, we can find the best place where to keep the patches.

File: media/css/firefox/features/article.scss (hunk 3 above)
Comment (janbrasna) [reply media/css/firefox/features/article.scss]: I've come to understand that any updates regarding nav (and breadcrumbs) in Protocol are now paused in expectations of a bigger redesign coming up, so probably if something of this is supposed to end up upstream, it's gonna be in the next design language anyways — and so with the nav due redesigning in a few months, I see this whole improvement as a tentative fix nonetheless, to be completely replaced with something more systematic, come summer(?)

For now I've managed to align the subnav & breadcrumbs perfectly to the main menu. But not sure I really love the outcome on some pages. 🤷

I also don't fancy the way I've patched up the overrides to implicit breadcrumb protocol styles here locally to with subnav 🤷

**Review by janbrasna — COMMENTED** [review]
So now this aligns exactly to the main nav bar, and while that improves the look&feel for pages with narrow content (think fx features, vpn… where the issue was originally raised), it also shows why it was aligned to the main content below in the first place, the most prominent pages that are wide and start with logos and headings look IMO better with the original alignment that treats the subnavs more as a content — than this "improved" alignment, which moves with the logo and main nav, not the content — depending on viewport these two start to align differently.

I'm taking some time to think about it a bit more, feel free to testdrive it locally, you'll soon notice what I mean…

Some random thoughts:

File: media/css/firefox/features/article.scss (hunk 3 above)
Comment (janbrasna) [reply media/css/firefox/features/article.scss]: @reemhamz Maybe the whole fix should be upstream https://github.com/mozilla/protocol/issues/933 after all — the ideal combo might be: subnav stays, only breadcrumbs get fixed (e.g. https://github.com/mozilla/bedrock/compare/main...8390523 but upstream instead), and what needs to change to better align with the rest is the main nav, not the other subnav elements (which are properly aligned to content in all breakpoints; it's the masthead that gets narrower sometimes…)

File: media/css/products/vpn/resource-center.scss [hunk 4]
```diff
@@ -82,6 +82,11 @@ $image-path: '/media/protocol/img';
     color: #666;
 }
 
+#outer-wrapper .mzp-c-breadcrumb {
+    // when stacked with subnav of the same color this makes it stand out
+    background: none;
```
[the remark this answers is not in the document]
Comment (reemhamz) [reply media/css/products/vpn/resource-center.scss:87]: I like this separation!

File: media/css/firefox/features/article.scss (hunk 3 above)
Comment (reemhamz) [reply media/css/firefox/features/article.scss]: Looking at your changes, I actually agree that the breadcrumb component should get fixed upstream in Protocol rather than have it override itself within Bedrock. However, I don't quite understand why the main nav needs to be changed? Could you further elaborate on that?

File: media/css/protocol/components/_sub-navigation.scss [hunk 5]
```diff
@@ -165,3 +175,24 @@
         }
     }
 }
+
+// * -------------------------------------------------------------------------- */
+// Overrides to align with subnav https://github.com/mozilla/bedrock/issues/14081
+
+    nav.mzp-c-breadcrumb {
```
[the remark this answers is not in the document]
Comment (reemhamz) [reply media/css/protocol/components/_sub-navigation.scss:182]: At this point, I believe it would make more sense to have these override style changes be a part of the Protocol component : )

Edit: Just viewed [your comment above](https://github.com/mozilla/bedrock/pull/14413#discussion_r1615283661) and looks like we agree

**Review by reemhamz — COMMENTED** [review]
Your changes are pretty good! However, I'm on the fence of adding them in Bedrock, and I think they would be better suited to be an update on the Breadcrumb component in Protocol instead.

File: media/css/protocol/components/_sub-navigation.scss [hunk 6]
```diff
@@ -13,12 +13,22 @@
         inset 0 10px 3px -10px rgba(29, 17, 51, 0.12);
 
     .mzp-l-content {
-        padding-top: 0;
-        padding-bottom: $spacing-md;
+        max-width: none;
```
Comment (stephaniehobson) [inline media/css/protocol/components/_sub-navigation.scss:16]: 😬  https://github.com/mozilla/protocol/issues/933

Comment (janbrasna) [reply media/css/protocol/components/_sub-navigation.scss:16]: @stephaniehobson Exactly, i'll most likely abandon all the changes here and start with the most important fix straight upstream https://github.com/mozilla/protocol/pull/938

File: media/css/firefox/features/article.scss (hunk 3 above)
Comment (janbrasna) [reply media/css/firefox/features/article.scss]: > I don't quite understand why the main nav needs to be changed? Could you further elaborate on that?

I'll post more info to the original issue #14081 to explain it a bit more, but just quickly following up your investigation from https://github.com/mozilla/bedrock/issues/14081#issuecomment-2054771820:

> I think I came to understand that the reason there's considered to be a misalignment is actually because the Breadcrumb component is aligned similarly to the subnav that we have. Now the question is, do we want to:
> 1. Change the alignments of both subnav and Breadcrumb component to align better with the main nav of the site
> 2. Only change the Breadcrumb component alignment
> 3. Keep things as is

Or there is 4.: Keep breadcrumbs and subnav aligned to mzp-l-content as it is now, but overcome the perception of these being misaligned in comparison to the main nav by actually tweaking some mq breakpoint styling of the nav instead, bc it's the main masthead that doesn't match mzp-l-content alignment in a handful of breakpoints. But I'll explain that a tad better with videos & examples so I'll keep you posted there…

TL;DR 3.) is out of the equation and https://github.com/mozilla/protocol/issues/933 needs fixing anyways, so 2.) will happen upstream https://github.com/mozilla/protocol/pull/938; 1.) was this PR and I don't like the result, so I will see about 4.) eventually, probably again in Protocol itself, but that's a separate issue then.
